src/stream_encrypt.rs

```diff
@@ -0,0 +1,1238 @@
+    /// # }
+    /// ```
+    pub fn new_from_memory(file_size: usize, chunk_store: F) -> Result<Self> {
+        if file_size < crate::MIN_ENCRYPTABLE_BYTES {
+            return Err(Error::Generic(format!(
+                "Too small for self-encryption! Required size at least {}",
+                crate::MIN_ENCRYPTABLE_BYTES
+            )));
+        }
+
+        let total_chunks = get_num_chunks(file_size);
+        if total_chunks < 3 {
+            return Err(Error::Generic(
+                "File must be large enough to generate at least 3 chunks".to_string(),
+            ));
+        }
+
+        Ok(Self {
+            file_size,
+            total_chunks,
+            chunk_store,
+            buffer: SparseBuffer::new(),
+            received_ranges: SparseBuffer::new(),
+            src_hashes: vec![None; total_chunks],
+            chunk_infos: vec![None; total_chunks],
+            chunks_src_processed: vec![false; total_chunks],
+            chunks_encrypted: vec![false; total_chunks],
+            deferred_chunks: vec![None; 2.min(total_chunks)],
+            is_complete: false,
+        })
+    }
+
+    /// Encrypts part of the data stream.
+    ///
+    /// This method processes incoming data, splits it into appropriate chunks according to
+    /// the pre-specified total file size, encrypts them, and stores them using the chunk_store
+    /// function. Returns `Some(DataMap)` only when the entire file encryption is complete.
+    ///
+    /// # Arguments
+    ///
+    /// * `start_pos` - The starting byte position of this data in the overall file
+    /// * `content` - The data content to encrypt (must align with expected position)
+    ///
+    /// # Returns
+    ///
+    /// * `Result<Option<DataMap>>` - None if more data is needed, Some(DataMap) when complete
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use self_encryption::StreamingEncrypt;
+    /// use bytes::Bytes;
+    /// use xor_name::XorName;
+    /// use std::{collections::HashMap, cell::RefCell, rc::Rc};
+    ///
+    /// # fn main() -> self_encryption::Result<()> {
+    /// let storage = Rc::new(RefCell::new(HashMap::new()));
+    /// let storage_clone = Rc::clone(&storage);
+    /// let store_chunk = move |hash: XorName, content: Bytes| -> self_encryption::Result<()> {
+    ///     let _ = storage_clone.borrow_mut().insert(hash, content.to_vec());
+    ///     Ok(())
+    /// };
+    ///
+    /// let file_size = 10_000;
+    /// let mut encryptor = StreamingEncrypt::new_from_memory(file_size, store_chunk)?;
+    ///
+    /// // Process first chunk of data
+    /// let data1 = Bytes::from(vec![1u8; 5000]);
+    /// let result1 = encryptor.encrypt_part(0, data1)?;
+    /// assert!(result1.is_none()); // Not complete yet
+    ///
+    /// // Process second chunk of data
+    /// let data2 = Bytes::from(vec![2u8; 5000]);
+    /// let result2 = encryptor.encrypt_part(5000, data2)?;
+    /// assert!(result2.is_some()); // Complete! Returns DataMap
```

Comment:
This is kind of hard to use, I feel like this could be done in one function call without the whole indexing process which is very error prone.